### PR TITLE
Addressed encoding issues for Blobs, Data Lake, and SMB Files

### PR DESCRIPTION
**Pull request by seanmcc-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/12738

Resolves https://github.com/Azure/azure-sdk-for-net/issues/12837

### CODE REVIEW

**Comment by seanmcc-msft** [comment]
Probably need to also check Data Lake and SMB Files for similar issue.

**Comment by seanmcc-msft** [comment]
TODO: add more comments around the UriBuilders and *Clients to make this change less ambiguous.

File: sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs [hunk 1]
```diff
@@ -301,7 +301,7 @@ private RequestUriBuilder BuildUri()
             if (!string.IsNullOrWhiteSpace(BlobContainerName))
             {
                 path.Append("/").Append(BlobContainerName);
-                if (!string.IsNullOrWhiteSpace(BlobName))
+                if (BlobName != null && BlobName.Length > 0)
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs:304]: is whitespace legit blob name ?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobUriBuilder.cs:304]: I believe so.

**Review by kasobol-msft — APPROVED** [review]

**Comment by seanmcc-msft** [comment]
The main goal of this PR - `BlobBaseClient.Name`, `DataLakeFileClient.Path` and `.Name`, `DataLakeDirectoryClient.Path` and `.Name`, `ShareFileClient.Path` and `.Name`, and `ShareDirectoryClient.Path` and `.Name` should be exactly what the user specified.  

If these fields contain special characters, they are encoded in the UriBuilders.  If we are generating these fields based off an Uri, we will unencode them once.

File: sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs [hunk 2]
```diff
@@ -84,40 +84,40 @@ public void FilePathsParsing()
             Uri uri1 = new Uri("http://dummyaccount.file.core.windows.net/share/dir1/dir2/file.txt");
             var builder1 = new ShareUriBuilder(uri1);
             var fileClient1 = new ShareFileClient(uri1);
-            TestHelper.AssertCacheableProperty("file.txt", () => fileClient1.Name);
-            TestHelper.AssertCacheableProperty("dir1/dir2/file.txt", () => fileClient1.Path);
+            Assert.AreEqual("file.txt", fileClient1.Name);
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs:87]: It weirds me out a little bit that I needed to change this.  Same with the other instances of `TestHelper.AssertCacheableProperty(...)` in `FileClientTests`.

**Comment by seanmcc-msft** [comment]
@tg-msft, @amnguye, could you take a look?

**Review by amnguye — APPROVED** [review]

**Pull request merged**